src/plugins/console/public/lib/osd/api.ts

```diff
@@ -44,41 +46,50 @@ import {
  * @constructor
  * @param bodyParametrizedComponentFactories same as urlParametrizedComponentFactories but used for body compilation
  */
-function Api(urlParametrizedComponentFactories, bodyParametrizedComponentFactories) {
-  this.globalRules = Object.create(null);
-  this.endpoints = Object.create(null);
-  this.urlPatternMatcher = new UrlPatternMatcher(urlParametrizedComponentFactories);
-  this.globalBodyComponentFactories = bodyParametrizedComponentFactories;
-  this.name = '';
-}
+export class Api {
+  globalRules: Record<string, any>;
+  endpoints: Record<string, Endpoint>;
+  urlPatternMatcher: UrlPatternMatcher;
+  globalBodyComponentFactories: ParametrizedComponentFactories | undefined;
+  name: string;
+
+  constructor(
+    urlParametrizedComponentFactories?: ParametrizedComponentFactories,
+    bodyParametrizedComponentFactories?: ParametrizedComponentFactories
+  ) {
+    this.globalRules = Object.create(null);
+    this.endpoints = Object.create(null);
+    this.urlPatternMatcher = new UrlPatternMatcher(urlParametrizedComponentFactories);
+    this.globalBodyComponentFactories = bodyParametrizedComponentFactories;
+    this.name = '';
+  }
 
-(function (cls) {
-  cls.addGlobalAutocompleteRules = function (parentNode, rules) {
+  addGlobalAutocompleteRules(parentNode: string, rules: Record<string, any>): void {
     this.globalRules[parentNode] = compileBodyDescription(
       'GLOBAL.' + parentNode,
       rules,
       this.globalBodyComponentFactories
     );
-  };
+  }
 
-  cls.getGlobalAutocompleteComponents = function (term, throwOnMissing) {
-    const result = this.globalRules[term];
+  getGlobalAutocompleteComponents(term: string, throwOnMissing?: boolean) {
```

Comment:
why only throwOnMissing has `?`

Reply:
I looked through the code and whenever this method used, the term was always present while throwOnMissing was optional.

Reply:
Here is the example, where throwOnMissing is optional
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/d94d7cb683aa76878bba5120cb8130add147d7ad/src/plugins/console/public/lib/autocomplete/autocomplete.ts#L883
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/d94d7cb683aa76878bba5120cb8130add147d7ad/src/plugins/console/public/lib/autocomplete/body_completer.ts#L127